File: math/code/binary-exponentiation/luogu-P1226-1.cpp
#include <cstdio>

// --8<-- [start:core]
long long binpow(long long a, long long b, long long p) {
  if (b == 0) return 1;
  long long res = binpow(a, b / 2, p);
  if (b % 2)
    return res * res % p * a % p;
  else
    return res * res % p;
}

// --8<-- [end:core]
int main() {
  int a, b, p;
  scanf("%d%d%d", &a, &b, &p);
  printf("%d^%d mod %d=%lld", a, b, p, binpow(a, b, p));
  return 0;
}

File: math/code/binary-exponentiation/luogu-P1226-1.py
# --8<-- [start:core]
def binpow(a, b, p):
    if b == 0:
        return 1
    res = binpow(a, b // 2, p)
    if (b % 2) == 1:
        return res * res * a % p
    else:
        return res * res % p


# --8<-- [end:core]
if __name__ == "__main__":
    a, b, p = map(int, input().split())
    print(f"{a}^{b} mod {p}={binpow(a, b, p)}")
